### DISCUSSION TITLE
Are there any types of modifications that *shouldn't* be submitted upstream?

### DISCUSSION THREAD

**Discussion by kitlith** [ppy/osu#13898] [Q&A]
There's obviously custom rulesets, that's not what i'm asking about.

It's been mentioned [here](https://github.com/ppy/osu/issues/9019#issuecomment-628467127) that mods should be merged into the repo, and that there's very little overhead preventing them from getting merged.

So what about something like this: https://github.com/hornyyy/Osu-Toy ? It doen't make sense as a custom ruleset to me (unless i'm just wrong >_>), my gut says it doesn't make sense to try to upstream it given its nsfw and dependency-adding nature, and is currently implemented as a custom mod (though frankly i think that if https://github.com/ppy/osu/issues/139 / https://github.com/ppy/osu/discussions/13011 ever got implemented it might be best done through that).

I have a whole idea based around taking advantage of the custom ruleset feature to get code loaded, then poke the game or use harmony to add custom mods to existing rulesets, as silly as that may seem due to this being an *open source project*, but distribution of said modification makes more sense like that than trying to distribute and keep up-to-date a custom build of osu!lazer, to me. But I figured that I should actually ask before i just sit down and try that at some point.

**Comment by peppy** [comment]
Using a custom ruleset to load code is heavily advised against, to the point we will actively be disallowing such rulesets when we get to the point of having control over it.

We *will* be exposing an external lighting API, which sounds like it could be used for your project. This is being tracked at https://github.com/ppy/osu/discussions/13011 as you mentioned, but is low priority so we'd be looking for someone to help implement.

It can probably be built on top of https://github.com/ppy/osu/pull/12274, which should be merged within the coming weeks.

**Reply by kitlith** [reply]
well, as long as whatever gets built on top of #12274 has a way to specify different types of data. it currently seems focused on information that changes infrequently, whereas the linked modification is more interested in *gameplay* information, like current health, score, combo.

regardless, while that does help for a particular type of modification, i am also trying to ask the more general question

> Using a custom ruleset to load code is heavily advised against, to the point we will actively be disallowing such rulesets when we get to the point of having control over it.

*shrug* you can replace the usage of custom ruleset with a proper 3rd party modloader and the question stays basically the same. will that be heavily discouraged? Will the "blessed" out-of-tree code only ever be custom rulesets, and the approved method for anything else is to distribute/maintain a fork?

**Reply by peppy** [reply]
"modloader" would be classified as a cheat in this case, if you're talking about being able to use it in an online context.

> Will the "blessed" out-of-tree code only ever be custom rulesets, and the approved method for anything else is to distribute/maintain a fork?

That's correct.
